cohortextractor/codelistlib.py

```diff
@@ -77,3 +77,31 @@ def combine_codelists(first_codelist, *other_codelists):
     return codelist(
         combined_dict.values(), first_codelist.system, check_categories=False
     )
+
+
+def expand_dmd_codelist(cl, vmp_mapping):
+    """Expand codelist to include all current and previous dm+d codes for VMPs.
+
+    VMP dm+d codes can change, and a codelist might contain a previous version of a VMP
+    code.  We attempt to handle this transparently by expanding the codelist to include
+    all current and previous dm+d codes for VMPs.  However, this means that a query
+    using `patients.with_these_medications(codelist, returning="code", ...)` might
+    return a code that is not in the provided codelist.
+    """
+
+    if cl.has_categories:
+        code_to_category = dict(cl)
+        for id, prev_id in vmp_mapping:
+            if id in code_to_category:
+                code_to_category[prev_id] = code_to_category[id]
+            if prev_id in code_to_category:
+                code_to_category[id] = code_to_category[prev_id]
+        return codelist(sorted(code_to_category.items()), cl.system)
+    else:
+        codes = set(cl)
+        for id, prev_id in vmp_mapping:
+            if id in codes:
+                codes.add(prev_id)
+            if prev_id in codes:
+                codes.add(id)
+        return codelist(sorted(codes), cl.system)
```

Comment:
if `TPPBackend.vmp_mapping` built a dict of `{id: set(other_ids)}` instead of a list of tuples, I think this could avoid checking every mapping tuple and just be:
```
codes_with_alternative_ids = codes & set(vmp_mapping)
for code in codes_with_alternative_ids:
    codes.update(vmp_mapping[code])
```
It would mean doing more work in the `TPPBackend.vmp_mapping` to make a dict that has all previous and subsequent ids for each code in the table, but since that's a cached property, maybe it's worth it?

In fact, I think it would just be one loop of the tuples there, rather than here for every codelist, i.e.:
```
mapping_dict = {}
for current, previous in vpm_mapping:
    mapping_dict.setdefault(current, set()).add(previous)
    mapping_dict.setdefault(previous, set()).add(current)
```

Reply:
Sorry, missed this when you wrote it.  It looks like a sensible suggestion, but I don't think there's any need to make this more efficient, as it's already much faster than talking to the database.